core/common/src/main/java/alluxio/resource/SharedResourcePool.java

```diff
@@ -0,0 +1,255 @@
+        // note: this is only partial order, but it works fine with
+        // priority queue
+        if (o1.equals(o2)) {
+          return 0;
+        }
+        if (o1.getRefCount() <= o2.getRefCount()) {
+          return -1;
+        }
+        return 1;
+      });
+
+  /** GC executor that closes unused resources periodically. */
+  private final ScheduledExecutorService mExecutor;
+
+  /** GC task future. */
+  private final ScheduledFuture<?> mGcFuture;
+
+  /**
+   * Create a new resource queue.
+   * @param options construction options
+   */
+  public SharedResourcePool(DynamicResourcePool.Options options) {
+    mExecutor = Preconditions.checkNotNull(options.getGcExecutor());
+    mMaxCapacity = options.getMaxCapacity();
+    mGcFuture = mExecutor.scheduleAtFixedRate(
+        () -> gcUnusedResources(true),
+        options.getInitialDelayMs(),
+        options.getGcIntervalMs(),
+        TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Acquire a {@link CloseableResource} from the queue, intending
+   * not to share.
+   *
+   * @return a closable resource
+   * @throws IOException when the resource pool is empty whereas we failed to create a new one
+   */
+  public CloseableResource<T> acquire() throws IOException {
+    return acquire(0);
+  }
+
+  /**
+   * Acquire a {@link CloseableResource} from the resource tree.
+   * This method accepts a best-effort based `maxAcceptableRefCount` parameter, which
+   * follows the following behavior in decreasing priority:
+   * 1. There is a resource in the tree that is current referenced less than
+   * `maxAcceptableRefCount`, then we reuse the resource.
+   * 2. If the current size of the resource pool is less than `maxCapacity`, then we
+   * try to create a new resource and use the new one.
+   * 3. If the resource pool is full or creating new one fails, then we fall back to sharing
+   * the least-referenced resource in the pool.
+   * 4. If the resource pool is empty and creating new resource fails, then we throw exception.
+   *
+   * @param maxAcceptableRefCount the maximum degree of acceptable sharing
+   * @return the closeable resource
+   * @throws IOException when the resource pool is empty whereas we failed to create a new one
+   */
+  public CloseableResource<T> acquire(int maxAcceptableRefCount) throws IOException {
+    try (LockResource r = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed);
+
+      // firstly try to reuse an eligible resource
+      CountingReference<T> ref = mResources.poll();
+      if (ref != null
+          && (ref.getRefCount() <= maxAcceptableRefCount
+              || mResources.size() + 1 >= mMaxCapacity)) {
+        ref.reference();
+        mResources.add(ref);
+        return makeResource(ref);
+      } else if (ref != null) {
+        mResources.add(ref);
+      }
+
+      // secondly try to create a new one
+      try {
+        CountingReference<T> newResource = new CountingReference<>(createNewResource(), 1);
```

Comment:
If `createNewResource` takes a long time, the outermost lock will block others who are willing to use a shared resource.

Reply:
It's a good point. My concern is that if `createNewResource` calls are not synchronized, we could end up exceed the `maxCapacity` badly in some high-concurrency situations.